src/accessTokenProvider.js

```diff
@@ -0,0 +1,111 @@
+import {
+  isResponseUnauthorized,
+} from './services/http';
+
+export class AccessTokenProvider {
+  constructor(fetch, config) {
+    this.fetch = fetch;
+
+    this.config = config;
+    this.refreshAccessTokenPromise = null;
+    this.tokens = {
+      refreshToken: null,
+      accessToken: null,
+    };
+
+    this.isAuthorized = this.isAuthorized.bind(this);
+    this.refresh = this.refresh.bind(this);
+    this.clear = this.clear.bind(this);
+
+    this.resolveAccessToken = this.resolveAccessToken.bind(this);
+    this.fetchToken = this.fetchToken.bind(this);
+    this.handleFetchResolved = this.handleFetchResolved.bind(this);
+    this.handleTokenResolved = this.handleTokenResolved.bind(this);
+    this.handleError = this.handleError.bind(this);
+  }
+
+  /**
+   * Refreshes current access token with provided refresh token
+   */
+  refresh() {
+    // if token resolver is not authorized it should just resolve
+    if (!this.isAuthorized()) {
+      return Promise.resolve();
+    }
+
+    // if we are not running token promise, start it
+    if (!this.refreshAccessTokenPromise) {
+      this.refreshAccessTokenPromise = new Promise(this.resolveAccessToken);
+    }
+
+    // otherwise just return existing promise
+    return this.refreshAccessTokenPromise;
+  }
+
+  /**
+   * Authorizes intercept library with given refresh token
+   * @param refreshToken
+   * @param accessToken
+   */
+  authorize(refreshToken, accessToken) {
+    this.tokens = { ...this.tokens, refreshToken, accessToken };
+  }
+
+  /**
+   * Returns current authorization for fetch fetchInterceptor
+   * @returns {{accessToken: string, refreshToken: string}}
+   */
+  getAuthorization() {
+    return this.tokens;
+  }
+
+  clear() {
+    this.tokens.accessToken = null;
+    this.tokens.refreshToken = null;
+  }
+
+  isAuthorized() {
+    return this.tokens.refreshToken !== null;
+  }
+
+  fetchToken(tokenRequest) {
+    const { fetch } = this;
+    return fetch(tokenRequest);
+  }
+
+  handleFetchResolved(response) {
+    this.refreshAccessTokenPromise = null;
+
+    if (isResponseUnauthorized(response)) {
+      this.clear();
+      return null;
```

Comment:
Same reasoning as on the comment above. This is not exceptional behavior because it's not network related, we've successfully fetched a response, and it's status causes provider to reset itself. Promises waiting for it to resolve will carry on without a new token and fail with 401 which is expected. The gist here is 'try to reauthorize to get new token, if it fails, it should fail the same manner as without fetch interceptor'.